port-forwarder/fwd_tcp.go

```diff
@@ -0,0 +1,230 @@
+}
+
+type PortForwardingOutgoingTcp struct {
+	l                     *logrus.Logger
+	tunService            *service.Service
+	cfg                   ForwardConfigOutgoingTcp
+	localListenConnection *net.TCPListener
+}
+
+func (fwd PortForwardingOutgoingTcp) Close() error {
+	fwd.localListenConnection.Close()
+	return nil
+}
+
+func (cf ForwardConfigOutgoingTcp) SetupPortForwarding(
+	tunService *service.Service,
+	l *logrus.Logger,
+) (io.Closer, error) {
+	localTcpListenAddr, err := net.ResolveTCPAddr("tcp", cf.localListen)
+	if err != nil {
+		return nil, err
+	}
+	remoteTcpAddr, err := net.ResolveTCPAddr("tcp", cf.remoteConnect)
+	if err != nil {
+		return nil, err
+	}
+	localListenPort, err := net.ListenTCP("tcp", localTcpListenAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("TCP port forwarding to '%v': listening on local TCP addr: '%v'",
+		remoteTcpAddr, localTcpListenAddr)
+
+	portForwarding := &PortForwardingOutgoingTcp{
+		l:                     l,
+		tunService:            tunService,
+		cfg:                   cf,
+		localListenConnection: localListenPort,
+	}
+
+	go portForwarding.acceptOnLocalListenPort()
+
+	return portForwarding, nil
+}
+
+func (pt *PortForwardingOutgoingTcp) acceptOnLocalListenPort() error {
+	for {
+		pt.l.Debug("listening on local TCP port ...")
+		connection, err := pt.localListenConnection.AcceptTCP()
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+
+		pt.l.Debugf("accept TCP connect from local TCP port: %v", connection.RemoteAddr())
+
+		go pt.handleClientConnection(connection)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection(localConnection *net.TCPConn) {
+	err := pt.handleClientConnection_intern(localConnection)
+	if err != nil {
+		pt.l.Debugf("Closed TCP client connection %s. Err: %v",
+			localConnection.LocalAddr().String(), err)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection_intern(localConnection net.Conn) error {
+
+	remoteConnection, err := pt.tunService.DialContext(context.Background(), "tcp", pt.cfg.remoteConnect)
+	if err != nil {
+		return err
+	}
+	return handleTcpClientConnection_generic(pt.l, localConnection, remoteConnection)
+}
+
+func handleTcpClientConnection_generic(l *logrus.Logger, connA, connB net.Conn) error {
+
+	dataTransferCtx, cancel := context.WithCancel(context.Background())
+
+	dataTransferHandler := func(from, to net.Conn) error {
+
+		name := fmt.Sprintf("%s -> %s", from.LocalAddr().String(), to.LocalAddr().String())
+
+		defer from.Close()
+		defer to.Close()
+		defer cancel()
+		// give communication in opposite direction time to finish as well
+		defer time.Sleep(time.Millisecond * 300)
+
+		// no write/read timeout
+		to.SetDeadline(time.Time{})
+		from.SetDeadline(time.Time{})
+		buf := make([]byte, 1*(1<<20))
+		for {
+			select {
+			case <-dataTransferCtx.Done():
+				return nil
+			default:
+			}
+
+			rn, r_err := from.Read(buf)
```

Comment:
I wonder if this could be replaced with `io.Copy()`

Reply:
it can. I once tried this. But it didn't improve the performance. Thats why I went on with other experiments. But now, as we solved the issue to a big extent, I will introduce it again as it simplifies the code. Thanks for pointing it out.

Reply:
I added a variant with io.Copy().
Problem: It seems to slightly decrease performance. So I'm wondering if I better should keep the original implementation.